YubiKit/YubiKit/Connections/Shared/Sessions/PIV/YKFPIVSession.m

```diff
@@ -857,6 +881,91 @@ - (void)blockPuk:(int)counter completion:(YKFPIVSessionGenericCompletionBlock)co
+    }
+    
+    NSData *temporaryPinData = nil;
+    if (!checkOnly) {
+        if (requestTemporaryPin) {
+            temporaryPinData = [[[YKFTLVRecord alloc] initWithTag:0x02 value:[NSData data]] data];
+        } else {
+            temporaryPinData = [[[YKFTLVRecord alloc] initWithTag:0x03 value:[NSData data]] data];
+        }
+    }
+    YKFAPDU *apdu = [[YKFAPDU alloc] initWithCla:0x00 ins: YKFPIVInsVerify p1:0x00 p2:YKFPIVSlotOCCAuth data:temporaryPinData type:YKFAPDUTypeExtended];
+    [self.smartCardInterface executeCommand:apdu completion:^(NSData * _Nullable data, NSError * _Nullable error) {
+        if (error) {
+            if (error.domain == YKFSessionErrorDomain && error.code == YKFAPDUErrorCodeReferencedDataNotFound) {
+                completion(nil, [[NSError alloc] initWithDomain:YKFPIVErrorDomain code:YKFPIVErrorCodeUnsupportedOperation userInfo:@{NSLocalizedDescriptionKey: @"Verify uv not supported by this YubiKey."}]);
+                return;
+            }
+            int retries = [self getRetriesFromStatusCode:(int)error.code];
+            if (retries >= 0) {
+                completion(nil, [YKFInvalidPinError invalidPinErrorWithRetries:retries]);
+                return;
+            } else {
+                completion(nil, error);
+                return;
+            }
+        }
+        if (requestTemporaryPin) {
+            completion(data, nil);
+        } else {
+            completion(nil, nil);
+        }
+    }];
+}
+
+- (void)verifyTemporaryPin:(NSData *)pinData completion:(nonnull YKFPIVSessionGenericCompletionBlock)completion {
+    if (pinData.length != 16) {
+        completion([[NSError alloc] initWithDomain:YKFPIVErrorDomain code:YKFPIVErrorCodeIllegalArgument userInfo:@{NSLocalizedDescriptionKey: @"Pin data length should be 16 bytes."}]);
+        return;
+    }
+    NSData *data = [[[YKFTLVRecord alloc] initWithTag:0x01 value:pinData] data];
+    YKFAPDU *apdu = [[YKFAPDU alloc] initWithCla:0 ins:YKFPIVInsVerify p1:0 p2:YKFPIVSlotOCCAuth data:data type:YKFAPDUTypeShort];
+    [self.smartCardInterface executeCommand:apdu completion:^(NSData * _Nullable data, NSError * _Nullable error) {
+        if (error == nil) {
+            completion(nil);
+            return;
+        } else {
+            if (error.domain == YKFSessionErrorDomain && error.code == YKFAPDUErrorCodeReferencedDataNotFound) {
+                completion([[NSError alloc] initWithDomain:YKFPIVErrorDomain code:YKFPIVErrorCodeUnsupportedOperation userInfo:@{NSLocalizedDescriptionKey: @"Verify uv not supported by this YubiKey."}]);
+                return;
+            }
+            int retries = [self getRetriesFromStatusCode:(int)error.code];
+            if (retries >= 0) {
+                completion([YKFInvalidPinError invalidPinErrorWithRetries:retries]);
+                return;
+            } else {
+                completion(error);
+                return;
+            }
```

Comment:
Review this part - if verification of the temporary PIN fails, the error.code will be 0x63C0, so there is no need to get the retries.